---
title: "Generation of DEG lists with revised methods"
output: html_notebook
---


```{r setup, include=FALSE}

library(openxlsx)
library(edgeR)
library(ggplot2)
library(gplots)
library(plotly)
library(stats)
library(VennDiagram)
library(stringr)
library(rprojroot)

rootDir <- find_root(is_rstudio_project)
dirPath <- file.path(rootDir, "results", "analysis_v3_output")
# old dirPath:  "C:/Users/Morgan/Documents/Lab Stuff/RNA seq/My_results/analysis_v3_output"

file.exists(dirPath)

```

## Description

### v3 Description Dec. 5, 2018

V3 of this analysis code is intended to be the final major revision, updates to this revision from v2 focus on naming comparisions and files more consistently, saving all results we need to a hierarchical folder system, and clean up of un-needed code chunks. 


### v2 notes
With Feedback from a meeting on Oct. 26, 2018, this analysis is updated to use more consistent parameters for defining sets of differentially expressed genes (DEGs).

The key updates are:
* Use more stringent filter for genes based on level of expression
* use wound control to eliminate genes before analysis to help FDR when identifying genes between S and R conditions
* Use a reasonable FDR cutouff
* use a consistent fold change filter

## Functions:

```{r}

makeFname <- function(folder, cont, change="", dtype="", FDRmax, logFCmin){
  filename <- paste(folder, "/", cont, change, dtype, "_FDR", 
                    str_pad(FDRmax*100, 2, pad="0"), "_LFC", 
                    str_pad(logFCmin,2, pad=0), "_", 
                    format(Sys.time(), "%Y%m%d"), sep="")
  if (dtype %in% c("", "_At_homologs")){
    filename <- paste(filename, ".csv", sep="")
  } else if (grepl("plot|hist|diag", dtype)){
    filename <- paste(filename, ".png", sep="")
  }
  return(filename)
}


save_DEGs <- function(DEGList, folder, cont, FDRmax, logFCmin, dir){
  DEGs <- DEGList
  DEGs$Gene_ID <- row.names(DEGs)
  DEGs <- dplyr::left_join(DEGs, edinburghGeneInfo, by="Gene_ID")
  row.names(DEGs) <- DEGs$Gene_ID
  
  # fname <- makeFname(folder=folder, cont=cont, FDRmax=FDRmax, logFCmin=logFCmin)
  # fPath <- file.path(dirPath, fname)
  # write.csv(DEGs, fPath, row.names=TRUE)
  
  fname <- makeFname(folder=folder, cont=cont, change="_UP", FDRmax=FDRmax, logFCmin=logFCmin)
  fPath <- file.path(dirPath, fname)
  write.csv(subset(DEGs, logFC > 0), fPath, row.names=TRUE)
  
  fname <- makeFname(folder=folder, cont=cont, change="_DOWN", FDRmax=FDRmax, logFCmin=logFCmin)
  fPath <- file.path(dirPath, fname)
  write.csv(subset(DEGs, logFC < 0), fPath, row.names=TRUE)
  
  fname <- makeFname(folder=paste(folder, "/GO_lists", sep=""), cont=cont, 
                     change="_UP", dtype="_At_homologs", FDRmax=FDRmax, logFCmin=logFCmin)
  fPath <- file.path(dirPath, fname)
  write.table(subset(DEGs, (logFC>0)&(!is.na(Symbol)), select="A._thaliana_best_hit"), 
              fPath, quote=FALSE, row.names=FALSE, na="", col.names=FALSE, sep=",")
  
  fname <- makeFname(folder=paste(folder, "/GO_lists", sep=""), cont=cont, 
                     change="_DOWN", dtype="_At_homologs", FDRmax=FDRmax, logFCmin=logFCmin)
  fPath <- file.path(dirPath, fname)
  write.table(subset(DEGs, (logFC<0)&(!is.na(Symbol)), select="A._thaliana_best_hit"), 
              fPath, quote=FALSE, row.names=FALSE, na="", col.names=FALSE, sep=",")
}



save_Venn_Lists <- function(allGenesList, folder, cont, FDRmax, logFCmin, change, dir){
  # AllGenesList should have edinInfo, and group columns 
  # change should be of form "_UP", or "_DOWN"
  
  # WT significant:
  fname <- makeFname(folder=folder, cont=paste(cont, "WT_SIG", sep="_"), 
                     change=change, FDRmax=FDRmax, logFCmin=logFCmin)
  fPath <- file.path(dirPath, fname)
  write.csv(subset(allGenesList, group=="WT_sig"), fPath, row.names=TRUE)
  
  fname <- makeFname(folder=paste(folder, "/GO_lists", sep=""), 
                     cont=paste(cont, "WT_SIG", sep="_"), 
                     change=change, dtype="_At_homologs", FDRmax=FDRmax, logFCmin=logFCmin)
  fPath <- file.path(dirPath, fname)
  write.table(subset(allGenesList, (group=="WT_sig")&(!is.na(Symbol)), 
                     select="A._thaliana_best_hit"), 
              fPath, quote=FALSE, row.names=FALSE, na="", col.names=FALSE, sep=",")
  
  # phyB significant:
  fname <- makeFname(folder=folder, cont=paste(cont, "phyB_SIG", sep="_"), 
                     change=change, FDRmax=FDRmax, logFCmin=logFCmin)
  fPath <- file.path(dirPath, fname)
  write.csv(subset(allGenesList, group=="phyB_sig"), fPath, row.names=TRUE)
  
  fname <- makeFname(folder=paste(folder, "/GO_lists", sep=""), 
                     cont=paste(cont, "phyB_SIG", sep="_"), 
                     change=change, dtype="_At_homologs", FDRmax=FDRmax, logFCmin=logFCmin)
  fPath <- file.path(dirPath, fname)
  write.table(subset(allGenesList, (group=="phyB_sig")&(!is.na(Symbol)), 
                     select="A._thaliana_best_hit"), 
              fPath, quote=FALSE, row.names=FALSE, na="", col.names=FALSE, sep=",")
  
  # both significant:
  fname <- makeFname(folder=folder, cont=paste(cont, "BOTH_SIG", sep="_"), 
                     change=change, FDRmax=FDRmax, logFCmin=logFCmin)
  fPath <- file.path(dirPath, fname)
  write.csv(subset(allGenesList, group=="Both_sig"), fPath, row.names=TRUE)
  
  fname <- makeFname(folder=paste(folder, "/GO_lists", sep=""), 
                     cont=paste(cont, "BOTH_SIG", sep="_"), 
                     change=change, dtype="_At_homologs", FDRmax=FDRmax, logFCmin=logFCmin)
  fPath <- file.path(dirPath, fname)
  write.table(subset(allGenesList, (group=="Both_sig")&(!is.na(Symbol)), 
                     select="A._thaliana_best_hit"), 
              fPath, quote=FALSE, row.names=FALSE, na="", col.names=FALSE, sep=",")
  
}




addWoundCtrl <- function(DEGs, woundIDs){
  output <- DEGs
  output$woundCtrl <- row.names(DEGs) %in% woundIDs
  return(output)
}


mergeGenes <- function(genesA, genesB){
  genesA$Gene_ID <- row.names(genesA)
  genesB$Gene_ID <- row.names(genesB)
  commonGenes <- dplyr::inner_join(genesA, genesB, by="Gene_ID")
  row.names(commonGenes) <- commonGenes$Gene_ID
  return(commonGenes)
}

se <- function(x){
  # calculate standard error of x
  sd(x)/sqrt(length(x))
} 

group2Line <- Vectorize(function(group){
  switch(group,
         phyB.24 = 1,
         phyB.P = 1,
         phyB.R = 2,
         phyB.S = 2,
         WT.24 = 3,
         WT.P = 3,
         WT.R = 4,
         WT.S = 4)
})

plotDE <- function(geneId, normExpr, display=TRUE, save=FALSE, saveDir="", fileType="svg", title=NULL, ...){
  normExpr <- data.frame(normExpr)
  group <- gsub("_\\d_", ".", colnames(normExpr))
  geneData <- as.data.frame(t(normExpr[geneId, ]))
  geneData$group <- group
  geneData <- dplyr::group_by(geneData, group)
  dataSummary <- dplyr::summarise(geneData, mean=mean(!!rlang::sym(geneId)), 
                                  SE=se(!!rlang::sym(geneId)))
  dataSummary <- tidyr::separate(dataSummary, group, 
                                 into=c("Genotype", "Timepoint"), sep="[.]", 
                                 remove=FALSE)
  dataSummary$Timepoint <- factor(dataSummary$Timepoint, levels=c("P", "24", "S", "R"))
  dataSummary$line <- group2Line(dataSummary$group)
  if (!is.null(title)){
    plotTitle <- title
  } else{
    plotTitle <- geneId
  }
    # make plot
  plot <- ggplot(dataSummary, aes(x=Timepoint, y=mean, color=Genotype)) + geom_point(size=2) + 
    geom_errorbar(aes(ymin=mean-SE, ymax=mean+SE), width=0.2) + geom_line(aes(group=line)) + 
    expand_limits(y=0) + labs(title=plotTitle, y="Relative Expression")
  if (display) {
    print(plot)
  }
  if (save) {
    if (!file.exists(saveDir)) {
      stop("the save directory does not exist")
    }
    saveFile <- file.path(saveDir, paste(geneId, " expression plot.", fileType, sep=""))
    if (file.exists(saveFile)) {
      warning(paste(saveFile, "is being overwritten"))
    }
    ggsave(saveFile, plot, device=fileType, ...)
  }
}


geneHeatmap <- function(genes, DGEdata, groupAv=TRUE, rsOnly=TRUE,
                        labels=NULL){
  
  DEGCpm <- cpm(DGEdata[genes, ])
  sampleOrder <- c(
    "WT_3_P",   "WT_4_P",   "WT_5_P",   "WT_6_P",    
    "phyB_3_P", "phyB_4_P", "phyB_5_P",
    "WT_3_24",  "WT_4_24",  "WT_5_24",  "WT_6_24", 
    "phyB_3_24","phyB_4_24","phyB_5_24",
    "WT_3_S",   "WT_4_S",   "WT_5_S",   "WT_6_S",    
    "phyB_3_S", "phyB_4_S", "phyB_5_S",
    "WT_3_R",   "WT_4_R",   "WT_5_R",   "WT_6_R",
    "phyB_3_R", "phyB_4_R", "phyB_5_R",
    "WT.S_P",   "phyB.S_P", "WT.R_S",   "phyB.R_S") 
  sampleOrder <- sampleOrder[sampleOrder %in% colnames(DEGCpm)]
  DEGCpm <- DEGCpm[, sampleOrder]
  group <- gsub("_\\d_", ".", colnames(DEGCpm))
  if (groupAv) {
    DEGCpm2 <- as.data.frame(t(DEGCpm))
    DEGCpm2 <- stats::aggregate(DEGCpm2, by=list(group), FUN=mean)
    row.names(DEGCpm2) <- DEGCpm2$Group.1
    DEGCpm2 <- DEGCpm2[, -1]
    DEGCpm <- t(DEGCpm2)
    groupOrder <- c("WT.P", "phyB.P", "WT.24", "phyB.24", "WT.S", "phyB.S", "WT.R", "phyB.R")
    groupOrder <- groupOrder[groupOrder %in% colnames(DEGCpm)]
    DEGCpm <- DEGCpm[, groupOrder]
  }
  if (rsOnly){
    DEGCpm <- DEGCpm[, !grepl("P|(24)", colnames(DEGCpm))]
  }
  if(!is.null(labels)) {
    hMapLabels <- labels    
  } else{
    hMapLabels <- row.names(DEGCpm)
  }
  hr <- hclust(as.dist(1-cor(t(DEGCpm), method="pearson")), method="complete")
  fig <- heatmap.2(DEGCpm, dendrogram="row", Rowv=as.dendrogram(hr), trace="none",
          col="bluered", scale="row", Colv=NA, margins=c(8,7),
          labRow=hMapLabels)
  return(fig)
}


commonGenes <- function(DEGsA, DEGsB){
  commonIDs <- row.names(DEGsA)[row.names(DEGsA) %in% row.names(DEGsB)]
  DEGsA[commonIDs, ]
}

doubleVenn <- function(DEG1, DEG2, name1="Set1", name2="Set2"){
  overlap <- nrow(commonGenes(DEG1, DEG2))
  
  draw.pairwise.venn(area1=nrow(DEG1), area2=nrow(DEG2), 
                     cross.area=overlap, 
                     fill=c("red", "blue"), category=c(name1, name2))
}

tripVenn <- function(DEG1, DEG2, DEG3, names=c("Set1", "Set2", "Set3")){
  genes12 <- commonGenes(DEG1, DEG2)
  genes23 <- commonGenes(DEG2, DEG3)  
  genes13 <- commonGenes(DEG1, DEG3)
  genes123 <- commonGenes(DEG1, genes23)
  
  draw.triple.venn(area1=nrow(DEG1), area2=nrow(DEG2), area3=nrow(DEG3),
                   n12=nrow(genes12), n23=nrow(genes23), n13=nrow(genes13), 
                   n123=nrow(genes123),
                   fill=c("red", "blue", "green"), category=names, scaled=FALSE)
}

nVenn <- function(DEGsList){
  DEGVenn <- Venn(DEGsList)
  gp <- VennThemes(compute.Venn(DEGVenn))
  gp$SetText <- lapply(gp$SetText,function(x) {x$fontsize<-10; return(x)})
  plot(DEGVenn, doWeights=FALSE,gp=gp)
}


setGroup <- function(row, change="up", sigLvl=0.01, LFCutoff=1){
  if (change=="up"){
    subsetMask <- (row$logFC_WT > LFCutoff | row$logFC_phyB > LFCutoff) & (row$FDR_WT < sigLvl | row$FDR_phyB < sigLvl)
  }
  if (change=="down"){
    subsetMask <- (row$FDR_WT < sigLvl | row$FDR_phyB < sigLvl) & (row$logFC_WT < -LFCutoff | row$logFC_phyB < -LFCutoff)
  }
  wtSigTest <-row$FDR_WT < sigLvl & row$FDR_phyB > sigLvl
  phyBSigTest <- row$FDR_phyB < sigLvl & row$FDR_WT > sigLvl
  group <- "none"
  if (subsetMask) {
    group <- "sig"
    if (wtSigTest) {
      group <- "WT_sig"
    }else if (phyBSigTest) {
      group <- "phyB_sig"
    } else {
      group <- "Both_sig"
    }
  }
  return(group)
}

addAvgCpm <- function(geneList, DGEdata, groups=NULL){
  
  genes <- row.names(geneList)
  
  DEGCpm <- cpm(DGEdata[genes, ])
  sampleOrder <- c(
    "WT_3_P",   "WT_4_P",   "WT_5_P",   "WT_6_P",    
    "phyB_3_P", "phyB_4_P", "phyB_5_P",
    "WT_3_24",  "WT_4_24",  "WT_5_24",  "WT_6_24", 
    "phyB_3_24","phyB_4_24","phyB_5_24",
    "WT_3_S",   "WT_4_S",   "WT_5_S",   "WT_6_S",    
    "phyB_3_S", "phyB_4_S", "phyB_5_S",
    "WT_3_R",   "WT_4_R",   "WT_5_R",   "WT_6_R",
    "phyB_3_R", "phyB_4_R", "phyB_5_R",
    "WT.S_P",   "phyB.S_P", "WT.R_S",   "phyB.R_S") 
  sampleOrder <- sampleOrder[sampleOrder %in% colnames(DEGCpm)]
  DEGCpm <- DEGCpm[, sampleOrder]

  if (class(DEGCpm)=="matrix") {
    group <- gsub("_\\d_", ".", colnames(DEGCpm)) 
    DEGCpm2 <- as.data.frame(t(DEGCpm))
  }else {
    group <- gsub("_\\d_", ".", names(DEGCpm)) 
    DEGCpm2 <- as.data.frame(DEGCpm)
    colnames(DEGCpm2) <- genes
  }
  DEGCpm2 <- stats::aggregate(DEGCpm2, by=list(group), FUN=mean)
  row.names(DEGCpm2) <- DEGCpm2$Group.1
  # DEGCpm2 <- DEGCpm2[, -1]
  DEGCpm <- data.frame(t(DEGCpm2))
  
  # row.names(DEGCpm) <- colnames(DEGCpm2)
  # colnames(DEGCpm) <- row.names(DEGCpm2)
  
  groupOrder <- c("WT.P", "phyB.P", "WT.24", "phyB.24", "WT.S", "phyB.S", "WT.R", "phyB.R")
  groupOrder <- groupOrder[groupOrder %in% colnames(DEGCpm)]
  DEGCpm <- DEGCpm[-1, groupOrder]
  if (!is.null(groups)){
    DEGCpm <- DEGCpm[, groups]
  }
  colnames(DEGCpm) <- paste(colnames(DEGCpm), "_Avg_CPM", sep="")
  return(cbind(geneList, DEGCpm))
  
}

```

## get gene Info from Edinburgh results

```{r}

edinburghPath <- file.path(rootDir, "data", "blast_top_hits.xlsx")
df <- read.xlsx(xlsxFile=edinburghPath, sheet=2, skipEmptyRows=FALSE)
colnames(df) <- gsub("(\\.)([a-z])", "_\\2", colnames(df))
colnames(df)[1] <- "Gene_ID"

edinburghGeneInfo <- df

```


## Loading the data

The code below for loading the data is largely unchanged from previous .Rmd docs.
With the exception that a more stringent filter is being used to filter out genes with low expression

The filter criteria is to remove rows(genes) that have less than 3 samples having above 100 counts.

```{r}

countsFile <- file.path(rootDir, "data", "raw_counts.csv")
# old path: "C:/Users/Morgan/Documents/Lab Stuff/RNA seq/raw_counts.csv"
counts <- read.csv(countsFile, row.names="X")

group <- factor(c(rep("WT.P",4), rep("WT.24", 4), rep("WT.S", 4), rep("WT.R", 4),
           rep("phyB.P", 3), rep("phyB.24", 3), rep("phyB.S", 3), 
           rep("phyB.R", 3)))

# create the DGEList object edgeR uses
DGEdata <- DGEList(counts=counts, group=group)

# filter out rows with fewer than 3 samples that have more than 10 counts/million
keep <- rowSums(cpm(DGEdata) > 10) >= 3
DGEdata <- DGEdata[keep, , keep.lib.sizes=FALSE]

# woundDGEdata <- DGEdata[, c(1:8,17:22)]
woundDGEdata <- DGEdata[, c(1:8,13:16,17:22,26:28)] # Excludes S timepoint

```

### add Edinburgh info to all genes
```{r}

allGeneIDs <- row.names(DGEdata)

fname <- file.path(dirPath, "all_gene_ids.csv")
write.table(allGeneIDs, fname, row.names=FALSE, na="", col.names=FALSE, sep=",")

allGenesEdin <- dplyr::left_join(data.frame(Gene_ID=allGeneIDs), 
                                 edinburghGeneInfo, by="Gene_ID")


# add the amaryllis file's info to this
fname <- file.path(rootDir, "data", "All genes and identities.xlsx")
df <- read.xlsx(xlsxFile=fname, sheet=1, skipEmptyRows=FALSE)

# add some genes from the amarullis file
dfNew <- df[!(df$gene.ID %in% edinburghGeneInfo$Gene_ID), c("gene.ID", "X9", "description")]
colnames(dfNew) <- c("Gene_ID", "A._thaliana_best_hit", "Description")
allGenesEdin <- data.table::rbindlist(list(allGenesEdin, dfNew), use.names=TRUE, fill=TRUE)


fname <- file.path(dirPath, "allGeneA.tHomologs.csv")
write.table(subset(allGenesEdin, select="A._thaliana_best_hit"), 
            fname, quote=FALSE, row.names=FALSE, na="", col.names=FALSE, sep=",")

```




 
### Modified design: ~gt + gt:trt  (no plant term)
```{r}

sampleCond <- data.frame(limma::strsplit2(colnames(counts), "_"))
colnames(sampleCond) <- c("Genotype", "Plant", "Timepoint")

# re order factor levels to ensure model intercept is correct
sampleCond$Genotype <- factor(sampleCond$Genotype, c("WT", "phyB"))
sampleCond$Plant <- factor(sampleCond$Plant, c("3", "4", "5", "6"))
sampleCond$Timepoint <- factor(sampleCond$Timepoint, c("P", "24", "S", "R"))

# create the model matrix used for the linear model
design <- model.matrix( ~Genotype + Genotype:Timepoint, data=sampleCond)
# remove column because there is no phyB plant #6
design <- design[, !(colnames(design) %in% c("GenotypephyB:Plant6"))] 

# simplify column names of design
colnames(design) <- gsub("Genotype|Timepoint|Plant", "", colnames(design))
colnames(design) <- gsub(":", ".", colnames(design))

# rename intercept column to avoid warning further down
colnames(design)[colnames(design) == "(Intercept)"] <- "Intercept" 
colnames(design)[colnames(design) == "24"] <- "T24"

```

### Modified design: ~gt + trt + gt:trt
```{r}
# 
# sampleCond <- data.frame(limma::strsplit2(colnames(counts), "_"))
# colnames(sampleCond) <- c("Genotype", "Plant", "Timepoint")
# 
# # re order factor levels to ensure model intercept is correct
# sampleCond$Genotype <- factor(sampleCond$Genotype, c("WT", "phyB"))
# sampleCond$Plant <- factor(sampleCond$Plant, c("3", "4", "5", "6"))
# sampleCond$Timepoint <- factor(sampleCond$Timepoint, c("P", "24", "S", "R"))
# 
# # create the model matrix used for the linear model
# design <- model.matrix( ~Genotype + Timepoint + Genotype:Timepoint, data=sampleCond)
# # remove column because there is no phyB plant #6
# design <- design[, !(colnames(design) %in% c("GenotypephyB:Plant6"))] 
# 
# # simplify column names of design
# colnames(design) <- gsub("Genotype|Timepoint|Plant", "", colnames(design))
# colnames(design) <- gsub(":", ".", colnames(design))
# 
# # rename intercept column to avoid warning further down
# colnames(design)[colnames(design) == "(Intercept)"] <- "Intercept" 
# colnames(design)[colnames(design) == "24"] <- "T24"

```



### design for calculating dispersion, account for gt:plant
```{r}

sampleCond <- data.frame(limma::strsplit2(colnames(counts), "_"))
colnames(sampleCond) <- c("Genotype", "Plant", "Timepoint")

# re order factor levels to ensure model intercept is correct
sampleCond$Genotype <- factor(sampleCond$Genotype, c("WT", "phyB"))
sampleCond$Plant <- factor(sampleCond$Plant, c("3", "4", "5", "6"))
sampleCond$Timepoint <- factor(sampleCond$Timepoint, c("P", "24", "S", "R"))

# create the model matrix used for the linear model
dispDesign <- model.matrix( ~Genotype + Genotype:Timepoint + Genotype:Plant, data=sampleCond)
# remove column because there is no phyB plant #6
dispDesign <- dispDesign[, !(colnames(dispDesign) %in% c("GenotypephyB:Plant6"))] 

# simplify column names of design
colnames(dispDesign) <- gsub("Genotype|Timepoint|Plant", "", colnames(dispDesign))
colnames(dispDesign) <- gsub(":", ".", colnames(dispDesign))

# rename intercept column to avoid warning further down
colnames(dispDesign)[colnames(dispDesign) == "(Intercept)"] <- "Intercept" 
colnames(dispDesign)[colnames(dispDesign) == "24"] <- "T24"

```



### Wound design:
```{r}

sampleCond <- data.frame(limma::strsplit2(colnames(woundDGEdata$counts), "_"))
colnames(sampleCond) <- c("Genotype", "Plant", "Timepoint")

# re order factor levels to ensure model intercept is correct
sampleCond$Genotype <- factor(sampleCond$Genotype, c("WT", "phyB"))
sampleCond$Plant <- factor(sampleCond$Plant, c("3", "4", "5", "6"))
sampleCond$Timepoint <- factor(sampleCond$Timepoint, c("P", "24", "R"))

# create the model matrix used for the linear model
woundDesign <- model.matrix( ~Genotype + Genotype:Timepoint + Genotype:Plant, data=sampleCond)
# remove column because there is no phyB plant #6
woundDesign <- woundDesign[, !(colnames(woundDesign) %in% c("GenotypephyB:Plant6"))] 

# simplify column names of design
colnames(woundDesign) <- gsub("Genotype|Timepoint|Plant", "", colnames(woundDesign))
colnames(woundDesign) <- gsub(":", ".", colnames(woundDesign))

# rename intercept column to avoid warning further down
colnames(woundDesign)[colnames(woundDesign) == "(Intercept)"] <- "Intercept" 
colnames(woundDesign)[colnames(woundDesign) == "24"] <- "T24"

```




### Contrasts

```{r}
fullContrasts <- makeContrasts("WT.24_vs_WT.P" = WT.24,
                               "phyB.24_vs_phyB.P" = phyB.24,
                               "WT.S_vs_WT.P" = WT.S,
                               "phyB.S_vs_phyB.P" = phyB.S,
                               "WT.R_vs_WT.S" = (WT.R - WT.S),  
                               "phyB.R_vs_phyB.S" = phyB.R - phyB.S,
                               "WT.R_vs_WT.P" = WT.R,
                               "phyB.R_vs_phyB.P" = phyB.R,
                               
                               
                               "phyB.24_vs_WT.24" = phyB.24 + phyB - WT.24,
                               "phyB.R_vs_WT.R" = phyB.R + phyB - WT.R,
                               "phyB.S_vs_WT.S" = phyB.S +phyB - WT.S,
                             
                               "phyB.R-S_vs_WT.R-S" = (phyB.R - phyB.S) - (WT.R - WT.S),
                               "phyB.S-P_vs_WT.S-P" = phyB.S - WT.S,
                               "phyB.R-P_vs_WT.R-P" = phyB.R - WT.R,
                               "phyB.24-P_vs_WT.24-P" = phyB.24 - WT.24,
                             
                               "phyB.P_vs_WT.P" = phyB,
                               
                               "phyB.PSR_vs_WT.PSR" = 3*phyB +  phyB.S + 
                                 phyB.R -  WT.S - WT.R,
                             
                             levels = design)
```

### Contrasts for gt + trt + gt:trt
```{r}
# fullContrasts <- makeContrasts("WT.24_vs_WT.P" = T24,
#                                "phyB.24_vs_phyB.P" = phyB.24 + T24,
#                                "WT.S_vs_WT.P" = S,
#                                "phyB.S_vs_phyB.P" = phyB.S + S,
#                                "WT.R_vs_WT.S" = R - S,  
#                                "phyB.R_vs_phyB.S" = phyB.R + R - phyB.S - S,
#                                "WT.R_vs_WT.P" = R,
#                                "phyB.R_vs_phyB.P" = phyB.R + R,
#                                
#                                
#                                "phyB.24_vs_WT.24" = phyB + phyB.24,
#                                "phyB.R_vs_WT.R" = phyB.R + phyB,
#                                "phyB.S_vs_WT.S" = phyB.S +phyB,
#                              
#                                "phyB.R-S_vs_WT.R-S" = (phyB.R + phyB) - (phyB.S + phyB),
#                                "phyB.S-P_vs_WT.S-P" = phyB.S + S - S,
#                                "phyB.R-P_vs_WT.R-P" = phyB.R + R - R,
#                                "phyB.24-P_vs_WT.24-P" = phyB.24 + T24 - T24,
#                              
#                                "phyB.P_vs_WT.P" = phyB,
#                                
#                                "phyB.PSR_vs_WT.PSR" = 3*phyB +  phyB.S + 
#                                  phyB.R,
#                              
#                              levels = design)
```




### linear model fit

```{r}
# woundDesign <- design[c(1:8,17:22), c(1:4)] 
# woundDGEdata <- DGEdata[, c(1:8,17:22)]
woundDGEdata <- calcNormFactors(woundDGEdata)
# estimate dispersion
woundDGEdata <- estimateDisp(woundDGEdata, design=woundDesign)
# generalized linear model
# woundFit <- glmQLFit(woundDGEdata, woundDesign)



DGEdata <- calcNormFactors(DGEdata)
# estimate dispersion
DGEdata <- estimateDisp(DGEdata, design=design)
# generalized linear model
LmFit <- glmQLFit(DGEdata, design)
```



# Analysis




## Wound Control

### phyB.(24-P) vs WT.(24-P)

```{r}
# --- inputs --- #
folder <- "wound_control"
cont <- "phyB.24-P_vs_WT.24-P"
conds <- c("WT.P", "WT.24", "phyB.P", "phyB.24")
FDRmax <- 0.05
logFCmin <- 1
# ---        --- #
qlf <- glmQLFTest(LmFit, contrast=fullContrasts[, cont])
allGenes <- as.data.frame(topTags(qlf, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addAvgCpm(DEGs, woundDGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)



fig <- hist(allGenes$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")

figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)

dev.copy(png, figFile)
dev.off


```



The linear model does not do a good job here, see how the p-value histogram drops off as it nears 0. we believe this is due to the dispersion/variance of the data being dominated by the S condition, removing S from the model improves the 

NOTE: even the exact test is influenced by the dispersion calculation and thus by the S condition.

To avoid our wound control gene lists being impacted by the S condition and the linear model, we will use direct comparisons with fischer exact tests to generate wound control genes.

### WT.24 vs WT.P

```{r echo=FALSE}
# --- inputs --- #
folder <- "wound_control"
cont <- "WT.24_vs_WT.P_exact"
conds <- c("WT.P", "WT.24")
FDRmax <- 0.05
logFCmin <- 1
# --- ------ --- #
exTest <- exactTest(woundDGEdata, pair=c("WT.24", "WT.P"))
allGenes <- as.data.frame(topTags(exTest, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addAvgCpm(DEGs, woundDGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)

woundDEGsWT <- DEGs

fig <- hist(allGenes$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")

figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)

dev.copy(png, figFile)
dev.off()

```


### phyB.24 - phyB.P 

```{r echo=FALSE}
# --- inputs --- #
folder <- "wound_control"
cont <- "phyB.24_vs_phyB.P_exact"
conds <- c("phyB.P", "phyB.24")
FDRmax <- 0.05
logFCmin <- 1
# --- ------ --- #
exTest <- exactTest(woundDGEdata, pair=c("phyB.24", "phyB.P"))
allGenes <- as.data.frame(topTags(exTest, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addAvgCpm(DEGs, woundDGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)

woundDEGsphyB <- DEGs

hist(subset(allGenes, !PValue==1)$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")

figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)

dev.copy(png, figFile)
dev.off()

```

### collection of all wound controll genes

```{r}
# --- inputs --- #
folder <- "wound_control"
cont <- "wound_ctrl_genes"
FDRmax <- 0.05
logFCmin <- 1
# --- ------ --- #
woundIDs <- unique(c(row.names(woundDEGsWT), row.names(woundDEGsphyB)))

fname <- makeFname(folder=folder, cont=cont, FDRmax=FDRmax, logFCmin=logFCmin)
fPath <- file.path(dirPath, fname)
write.table(woundIDs, fPath, row.names=FALSE, na="", col.names=FALSE, sep=",")

```






# Timepoint v. Timepoint

## S vs P

### WT S vs P

```{r}
# --- inputs --- #
folder <- "timepoint_v_timepoint"
cont <- "WT.S_vs_WT.P"
conds <- c("WT.P", "WT.S")
FDRmax <- 0.01
logFCmin <- 1
# --- ------ --- #

qlf <- glmQLFTest(LmFit, contrast=fullContrasts[, cont])
allGenes <- as.data.frame(topTags(qlf, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addWoundCtrl(DEGs, woundIDs)
DEGs <- addAvgCpm(DEGs, DGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)

hist(allGenes$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")
figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()

# -- specific --- # 
allGenesWT_SvP <- allGenes
DEGsWT_SvP <- DEGs

head(DEGs, 20)

```


### phyB S vs P

```{r}
# --- inputs --- #
folder <- "timepoint_v_timepoint"
cont <- "phyB.S_vs_phyB.P"
conds <- c("phyB.P", "phyB.S")
FDRmax <- 0.01
logFCmin <- 1
# --- ------ --- #

qlf <- glmQLFTest(LmFit, contrast=fullContrasts[, cont])
allGenes <- as.data.frame(topTags(qlf, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addWoundCtrl(DEGs, woundIDs)
DEGs <- addAvgCpm(DEGs, DGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)

hist(allGenes$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")
figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()

# -- specific --- # 
allGenesphyB_SvP <- allGenes
DEGsphyB_SvP <- DEGs

head(DEGs, 20)

```


### phyB S-P vs WT S-P

```{r}
# --- inputs --- #
folder <- "geno_x_timepoint"
cont <- "phyB.S-P_vs_WT.S-P"
conds <- c("WT.P", "phyB.P", "WT.S", "phyB.S")
FDRmax <- 0.1
logFCmin <- 0
# --- ------ --- #

qlf <- glmQLFTest(LmFit, contrast=fullContrasts[, cont])
allGenes <- as.data.frame(topTags(qlf, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addWoundCtrl(DEGs, woundIDs)
DEGs <- addAvgCpm(DEGs, DGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)

hist(allGenes$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")
figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()

# --- specific --- #

DEGs_SvP <- DEGs
```


### Venn diagram of S vs. P

```{r}
# --- inputs --- #
folder <- "timepoint_v_timepoint/SvP_venn"
cont <- "SvP"
conds <- c("WT.P", "phyB.P", "WT.S", "phyB.S")
FDRmax <- 0.01
logFCmin <- 1
# --- ------ --- #


allGenes_SvP <- mergeGenes(allGenesWT_SvP, allGenesphyB_SvP)
allGenes_SvP <- addAvgCpm(allGenes_SvP, DGEdata, groups=conds)
colnames(allGenes_SvP) <- gsub("\\.x", "_WT", colnames(allGenes_SvP))
colnames(allGenes_SvP) <- gsub("\\.y", "_phyB", colnames(allGenes_SvP))



allGenes_SvP <- dplyr::left_join(allGenes_SvP, edinburghGeneInfo, by="Gene_ID")
row.names(allGenes_SvP) <- allGenes_SvP$Gene_ID

fname <- makeFname(folder=folder, cont="SvP_all_genes", FDRmax=FDRmax, logFCmin=logFCmin)
fPath <- file.path(dirPath, fname)
write.csv(allGenes_SvP, fPath, row.names=TRUE)

# --- UP --- #
group <- plyr::aaply(.data=allGenes_SvP, .margins=1, .fun=setGroup, change="up",
                       sigLvl=FDRmax, LFCutoff=logFCmin, .expand=FALSE)
allGenes_SvP$group <- group

save_Venn_Lists(allGenesList=allGenes_SvP, 
                folder=folder,
                cont=cont,
                FDRmax=FDRmax,
                logFCmin=logFCmin,
                change="_UP",
                dir=dirPath)

sum(allGenes_SvP$group %in% c("WT_sig", "Both_sig"))


plot.new()
temp <- draw.pairwise.venn(area1=sum(allGenes_SvP$group %in% c("WT_sig", "Both_sig")), 
                           area2=sum(allGenes_SvP$group %in% c("phyB_sig", "Both_sig")), 
                           cross.area=sum(allGenes_SvP$group %in% c("Both_sig")), 
                   fill=c("blue", "red"), category=c("WT S vs. P", "phyB S vs. P"))
grid.draw(temp)

figFile <- makeFname(folder=folder, cont=paste(cont, "_UP", sep=""), dtype="_Venn_diag", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()


cols <- c("none"="grey50", "phyB_sig"="red", "WT_sig"="blue", "Both_sig"="purple")
plotData <- allGenes_SvP
ggplot(plotData, aes(x=logFC_WT, y=logFC_phyB, color=group)) +
  geom_point(data=subset(plotData, group=="none"), alpha=0.2, stroke=0) +
  geom_point(data=subset(plotData, group=="Both_sig"), alpha=0.5, stroke=0) +
  geom_point(data=subset(plotData, group=="phyB_sig"), alpha=0.5, stroke=0) +
  geom_point(data=subset(plotData, group=="WT_sig"), alpha=0.5, stroke=0) +
  scale_color_manual(values=cols) +
  labs(x = "WT logFC", y="phyB logFC", title="logFC S vs. P")

figFile <- makeFname(folder=folder, cont=paste(cont, "_UP", sep=""), dtype="_LogFC_plot", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()

```

```{r}
# --- DOWN --- #
group <- plyr::aaply(.data=allGenes_SvP, .margins=1, .fun=setGroup, change="down",
                       sigLvl=FDRmax, LFCutoff=logFCmin, .expand=FALSE)
allGenes_SvP$group <- group

save_Venn_Lists(allGenesList=allGenes_SvP, 
                folder=folder,
                cont=cont,
                FDRmax=FDRmax,
                logFCmin=logFCmin,
                change="_DOWN",
                dir=dirPath)

sum(allGenes_SvP$group %in% c("WT_sig", "Both_sig"))


plot.new()
temp <- draw.pairwise.venn(area1=sum(allGenes_SvP$group %in% c("WT_sig", "Both_sig")), 
                           area2=sum(allGenes_SvP$group %in% c("phyB_sig", "Both_sig")), 
                           cross.area=sum(allGenes_SvP$group %in% c("Both_sig")), 
                   fill=c("blue", "red"), category=c("WT S vs. P", "phyB S vs. P"))
grid.draw(temp)

figFile <- makeFname(folder=folder, cont=paste(cont, "_DOWN", sep=""), dtype="_Venn_diag", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()


cols <- c("none"="grey50", "phyB_sig"="red", "WT_sig"="blue", "Both_sig"="purple")
plotData <- allGenes_SvP
ggplot(plotData, aes(x=logFC_WT, y=logFC_phyB, color=group)) +
  geom_point(data=subset(plotData, group=="none"), alpha=0.2, stroke=0) +
  geom_point(data=subset(plotData, group=="Both_sig"), alpha=0.5, stroke=0) +
  geom_point(data=subset(plotData, group=="phyB_sig"), alpha=0.5, stroke=0) +
  geom_point(data=subset(plotData, group=="WT_sig"), alpha=0.5, stroke=0) +
  scale_color_manual(values=cols) +
  labs(x = "WT logFC", y="phyB logFC", title="logFC S vs. P")

figFile <- makeFname(folder=folder, cont=paste(cont, "_DOWN", sep=""), dtype="_LogFC_plot", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()
```







## R vs. S


### WT R vs. S

```{r}
# --- inputs --- #
folder <- "timepoint_v_timepoint"
cont <- "WT.R_vs_WT.S"
conds <- c("WT.S", "WT.R")
FDRmax <- 0.01
logFCmin <- 1
# --- ------ --- #

qlf <- glmQLFTest(LmFit, contrast=fullContrasts[, cont])
allGenes <- as.data.frame(topTags(qlf, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addWoundCtrl(DEGs, woundIDs)
DEGs <- addAvgCpm(DEGs, DGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)

hist(allGenes$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")
figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()

# -- specific --- # 
allGenesWT_RvS <- allGenes
DEGsWT_RvS <- DEGs

head(DEGs, 20)

```


### phyB R vs. S

```{r}
# --- inputs --- #
folder <- "timepoint_v_timepoint"
cont <- "phyB.R_vs_phyB.S"
conds <- c("phyB.S", "phyB.R")
FDRmax <- 0.01
logFCmin <- 1
# --- ------ --- #

qlf <- glmQLFTest(LmFit, contrast=fullContrasts[, cont])
allGenes <- as.data.frame(topTags(qlf, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addWoundCtrl(DEGs, woundIDs)
DEGs <- addAvgCpm(DEGs, DGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)

hist(allGenes$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")
figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()

# -- specific --- # 
allGenesphyB_RvS <- allGenes
DEGsphyB_RvS <- DEGs

head(DEGs, 20)

```


### phyB R-S vs WT R-S

```{r}
# --- inputs --- #
folder <- "geno_x_timepoint"
cont <- "phyB.R-S_vs_WT.R-S"
conds <- c("WT.S", "phyB.S", "WT.R", "phyB.R")
FDRmax <- 0.1
logFCmin <- 0
# --- ------ --- #

qlf <- glmQLFTest(LmFit, contrast=fullContrasts[, cont])
allGenes <- as.data.frame(topTags(qlf, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addWoundCtrl(DEGs, woundIDs)
DEGs <- addAvgCpm(DEGs, DGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)

hist(allGenes$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")
figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()

# --- specific --- #

DEGs_RvS <- DEGs
```



### venn diagarm R vs. S
```{r}
# --- inputs --- #
folder <- "timepoint_v_timepoint/RvS_venn"
cont <- "RvS"
conds <- c("WT.S", "phyB.S", "WT.R", "phyB.R")
FDRmax <- 0.01
logFCmin <- 1
# --- ------ --- #


allGenes_RvS <- mergeGenes(allGenesWT_RvS, allGenesphyB_RvS)
allGenes_RvS <- addAvgCpm(allGenes_RvS, DGEdata, groups=conds)
colnames(allGenes_RvS) <- gsub("\\.x", "_WT", colnames(allGenes_RvS))
colnames(allGenes_RvS) <- gsub("\\.y", "_phyB", colnames(allGenes_RvS))



allGenes_RvS <- dplyr::left_join(allGenes_RvS, edinburghGeneInfo, by="Gene_ID")
row.names(allGenes_RvS) <- allGenes_RvS$Gene_ID

fname <- makeFname(folder=folder, cont="RvS_all_genes", FDRmax=FDRmax, logFCmin=logFCmin)
fPath <- file.path(dirPath, fname)
write.csv(allGenes_RvS, fPath, row.names=TRUE)

# --- UP --- #
group <- plyr::aaply(.data=allGenes_RvS, .margins=1, .fun=setGroup, change="up",
                       sigLvl=FDRmax, LFCutoff=logFCmin, .expand=FALSE)
allGenes_RvS$group <- group

save_Venn_Lists(allGenesList=allGenes_RvS, 
                folder=folder,
                cont=cont,
                FDRmax=FDRmax,
                logFCmin=logFCmin,
                change="_UP",
                dir=dirPath)

sum(allGenes_RvS$group %in% c("WT_sig", "Both_sig"))


plot.new()
temp <- draw.pairwise.venn(area1=sum(allGenes_RvS$group %in% c("WT_sig", "Both_sig")), 
                           area2=sum(allGenes_RvS$group %in% c("phyB_sig", "Both_sig")), 
                           cross.area=sum(allGenes_RvS$group %in% c("Both_sig")), 
                   fill=c("blue", "red"), category=c("WT R vs. S", "phyB R vs. S"))
grid.draw(temp)

figFile <- makeFname(folder=folder, cont=paste(cont, "_UP", sep=""), dtype="_Venn_diag", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()

cols <- c("none"="grey50", "phyB_sig"="red", "WT_sig"="blue", "Both_sig"="purple")
plotData <- allGenes_RvS
ggplot(plotData, aes(x=logFC_WT, y=logFC_phyB, color=group)) +
  geom_point(data=subset(plotData, group=="none"), alpha=0.2, stroke=0) +
  geom_point(data=subset(plotData, group=="Both_sig"), alpha=0.5, stroke=0) +
  geom_point(data=subset(plotData, group=="phyB_sig"), alpha=0.5, stroke=0) +
  geom_point(data=subset(plotData, group=="WT_sig"), alpha=0.5, stroke=0) +
  scale_color_manual(values=cols) +
  labs(x = "WT logFC", y="phyB logFC", title="logFC R vs. S")

figFile <- makeFname(folder=folder, cont=paste(cont, "_UP", sep=""), dtype="_LogFC_plot", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()


```

```{r}
# --- DOWN --- #
group <- plyr::aaply(.data=allGenes_RvS, .margins=1, .fun=setGroup, change="down",
                       sigLvl=FDRmax, LFCutoff=logFCmin, .expand=FALSE)
allGenes_RvS$group <- group

save_Venn_Lists(allGenesList=allGenes_RvS, 
                folder=folder,
                cont=cont,
                FDRmax=FDRmax,
                logFCmin=logFCmin,
                change="_DOWN",
                dir=dirPath)

sum(allGenes_RvS$group %in% c("WT_sig", "Both_sig"))


plot.new()
temp <- draw.pairwise.venn(area1=sum(allGenes_RvS$group %in% c("WT_sig", "Both_sig")), 
                           area2=sum(allGenes_RvS$group %in% c("phyB_sig", "Both_sig")), 
                           cross.area=sum(allGenes_RvS$group %in% c("Both_sig")), 
                   fill=c("blue", "red"), category=c("WT R vs. S", "phyB R vs. S"))
grid.draw(temp)

figFile <- makeFname(folder=folder, cont=paste(cont, "_DOWN", sep=""), dtype="_Venn_diag", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()



cols <- c("none"="grey50", "phyB_sig"="red", "WT_sig"="blue", "Both_sig"="purple")
plotData <- allGenes_RvS
ggplot(plotData, aes(x=logFC_WT, y=logFC_phyB, color=group)) +
  geom_point(data=subset(plotData, group=="none"), alpha=0.2, stroke=0) +
  geom_point(data=subset(plotData, group=="Both_sig"), alpha=0.5, stroke=0) +
  geom_point(data=subset(plotData, group=="phyB_sig"), alpha=0.5, stroke=0) +
  geom_point(data=subset(plotData, group=="WT_sig"), alpha=0.5, stroke=0) +
  scale_color_manual(values=cols) +
  labs(x = "WT logFC", y="phyB logFC", title="logFC R vs. S")

figFile <- makeFname(folder=folder, cont=paste(cont, "_DOWN", sep=""), dtype="_LogFC_plot", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()
```


## R vs. P


### WT R vs. P

```{r}
# --- inputs --- #
folder <- "timepoint_v_timepoint"
cont <- "WT.R_vs_WT.P"
conds <- c("WT.P", "WT.R")
FDRmax <- 0.01
logFCmin <- 1
# --- ------ --- #

qlf <- glmQLFTest(LmFit, contrast=fullContrasts[, cont])
allGenes <- as.data.frame(topTags(qlf, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addWoundCtrl(DEGs, woundIDs)
DEGs <- addAvgCpm(DEGs, DGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)

hist(allGenes$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")
figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()

# -- specific --- # 
allGenesWT_RvP <- allGenes
DEGsWT_RvP <- DEGs

head(DEGs, 20)

```


### phyB R vs. P

```{r}
# --- inputs --- #
folder <- "timepoint_v_timepoint"
cont <- "phyB.R_vs_phyB.P"
conds <- c("phyB.P", "phyB.R")
FDRmax <- 0.01
logFCmin <- 1
# --- ------ --- #

qlf <- glmQLFTest(LmFit, contrast=fullContrasts[, cont])
allGenes <- as.data.frame(topTags(qlf, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addWoundCtrl(DEGs, woundIDs)
DEGs <- addAvgCpm(DEGs, DGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)

hist(allGenes$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")
figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()

# -- specific --- # 
allGenesphyB_RvP <- allGenes
DEGsphyB_RvP <- DEGs

head(DEGs, 20)

```


### phyB R-P vs WT R-P

```{r}
# --- inputs --- #
folder <- "geno_x_timepoint"
cont <- "phyB.R-P_vs_WT.R-P"
conds <- c("WT.P", "phyB.P", "WT.R", "phyB.R")
FDRmax <- 0.1
logFCmin <- 0
# --- ------ --- #

qlf <- glmQLFTest(LmFit, contrast=fullContrasts[, cont])
allGenes <- as.data.frame(topTags(qlf, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addWoundCtrl(DEGs, woundIDs)
DEGs <- addAvgCpm(DEGs, DGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)

hist(allGenes$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")
figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()

# --- specific --- #

DEGs_RvP <- DEGs
```



### venn diagarm R vs. P

```{r}
# --- inputs --- #
folder <- "timepoint_v_timepoint/RvP_venn"
cont <- "RvP"
conds <- c("WT.P", "phyB.P", "WT.R", "phyB.R")
FDRmax <- 0.01
logFCmin <- 1
# --- ------ --- #


allGenes_RvP <- mergeGenes(allGenesWT_RvP, allGenesphyB_RvP)
allGenes_RvP <- addAvgCpm(allGenes_RvP, DGEdata, groups=conds)
colnames(allGenes_RvP) <- gsub("\\.x", "_WT", colnames(allGenes_RvP))
colnames(allGenes_RvP) <- gsub("\\.y", "_phyB", colnames(allGenes_RvP))



allGenes_RvP <- dplyr::left_join(allGenes_RvP, edinburghGeneInfo, by="Gene_ID")
row.names(allGenes_RvP) <- allGenes_RvP$Gene_ID

fname <- makeFname(folder=folder, cont="RvP_all_genes", FDRmax=FDRmax, logFCmin=logFCmin)
fPath <- file.path(dirPath, fname)
write.csv(allGenes_RvP, fPath, row.names=TRUE)

# --- UP --- #
group <- plyr::aaply(.data=allGenes_RvP, .margins=1, .fun=setGroup, change="up",
                       sigLvl=FDRmax, LFCutoff=logFCmin, .expand=FALSE)
allGenes_RvP$group <- group

save_Venn_Lists(allGenesList=allGenes_RvP, 
                folder=folder,
                cont=cont,
                FDRmax=FDRmax,
                logFCmin=logFCmin,
                change="_UP",
                dir=dirPath)

sum(allGenes_RvP$group %in% c("WT_sig", "Both_sig"))


plot.new()
temp <- draw.pairwise.venn(area1=sum(allGenes_RvP$group %in% c("WT_sig", "Both_sig")), 
                           area2=sum(allGenes_RvP$group %in% c("phyB_sig", "Both_sig")), 
                           cross.area=sum(allGenes_RvP$group %in% c("Both_sig")), 
                   fill=c("blue", "red"), category=c("WT R vs. P", "phyB R vs. P"))
grid.draw(temp)

figFile <- makeFname(folder=folder, cont=paste(cont, "_UP", sep=""), dtype="_Venn_diag", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()


cols <- c("none"="grey50", "phyB_sig"="red", "WT_sig"="blue", "Both_sig"="purple")
plotData <- allGenes_RvP
ggplot(plotData, aes(x=logFC_WT, y=logFC_phyB, color=group)) +
  geom_point(data=subset(plotData, group=="none"), alpha=0.2, stroke=0) +
  geom_point(data=subset(plotData, group=="Both_sig"), alpha=0.5, stroke=0) +
  geom_point(data=subset(plotData, group=="phyB_sig"), alpha=0.5, stroke=0) +
  geom_point(data=subset(plotData, group=="WT_sig"), alpha=0.5, stroke=0) +
  scale_color_manual(values=cols) +
  labs(x = "WT logFC", y="phyB logFC", title="logFC R vs. P")

figFile <- makeFname(folder=folder, cont=paste(cont, "_UP", sep=""), dtype="_LogFC_plot", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()

```

```{r}
# --- DOWN --- #
group <- plyr::aaply(.data=allGenes_RvP, .margins=1, .fun=setGroup, change="down",
                       sigLvl=FDRmax, LFCutoff=logFCmin, .expand=FALSE)
allGenes_RvP$group <- group

save_Venn_Lists(allGenesList=allGenes_RvP, 
                folder=folder,
                cont=cont,
                FDRmax=FDRmax,
                logFCmin=logFCmin,
                change="_DOWN",
                dir=dirPath)

sum(allGenes_RvP$group %in% c("WT_sig", "Both_sig"))


plot.new()
temp <- draw.pairwise.venn(area1=sum(allGenes_RvP$group %in% c("WT_sig", "Both_sig")), 
                           area2=sum(allGenes_RvP$group %in% c("phyB_sig", "Both_sig")), 
                           cross.area=sum(allGenes_RvP$group %in% c("Both_sig")), 
                   fill=c("blue", "red"), category=c("WT R vs. P", "phyB R vs. P"))
grid.draw(temp)

figFile <- makeFname(folder=folder, cont=paste(cont, "_DOWN", sep=""), dtype="_Venn_diag", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()

cols <- c("none"="grey50", "phyB_sig"="red", "WT_sig"="blue", "Both_sig"="purple")
plotData <- allGenes_RvP
ggplot(plotData, aes(x=logFC_WT, y=logFC_phyB, color=group)) +
  geom_point(data=subset(plotData, group=="none"), alpha=0.2, stroke=0) +
  geom_point(data=subset(plotData, group=="Both_sig"), alpha=0.5, stroke=0) +
  geom_point(data=subset(plotData, group=="phyB_sig"), alpha=0.5, stroke=0) +
  geom_point(data=subset(plotData, group=="WT_sig"), alpha=0.5, stroke=0) +
  scale_color_manual(values=cols) +
  labs(x = "WT logFC", y="phyB logFC", title="logFC R vs. P")

figFile <- makeFname(folder=folder, cont=paste(cont, "_DOWN", sep=""), dtype="_LogFC_plot", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()
```





## phyB vs WT at each timepoint (exact tests)


### phyB.P vs WT.P

Note: we're using woundDGEdata here

```{r echo=FALSE}
# --- inputs --- #
folder <- "phyB_v_WT_at_timepoint"
cont <- "phyB.P_vs_WT.P_exact"
conds <- c("phyB.P", "WT.P")
FDRmax <- 0.05
logFCmin <- 1
# --- ------ --- #
exTest <- exactTest(woundDGEdata, pair=conds)
allGenes <- as.data.frame(topTags(exTest, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addWoundCtrl(DEGs, woundIDs)
DEGs <- addAvgCpm(DEGs, DGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)

hist(allGenes$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")

figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)

dev.copy(png, figFile)
dev.off()

# --- specific --- #
allGenesP <- allGenes
DEGsP <- DEGs
```


### phyB.24 vs WT.24

Note: we're using woundDGEdata here

```{r echo=FALSE}
# --- inputs --- #
folder <- "phyB_v_WT_at_timepoint"
cont <- "phyB.24_vs_WT.24_exact"
conds <- c("phyB.24", "WT.24")
FDRmax <- 0.05
logFCmin <- 1
# --- ------ --- #
exTest <- exactTest(woundDGEdata, pair=conds)
allGenes <- as.data.frame(topTags(exTest, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addWoundCtrl(DEGs, woundIDs)
DEGs <- addAvgCpm(DEGs, DGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)

hist(allGenes$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")

figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)

dev.copy(png, figFile)
dev.off()

# --- specific --- #
allGenes24 <- allGenes
DEGs24 <- DEGs
```


### phyB.S vs WT.S

```{r echo=FALSE}
# --- inputs --- #
folder <- "phyB_v_WT_at_timepoint"
cont <- "phyB.S_vs_WT.S_exact"
conds <- c("phyB.S", "WT.S")
FDRmax <- 0.05
logFCmin <- 1
# --- ------ --- #
exTest <- exactTest(DGEdata, pair=conds)
allGenes <- as.data.frame(topTags(exTest, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addWoundCtrl(DEGs, woundIDs)
DEGs <- addAvgCpm(DEGs, DGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)

hist(allGenes$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")

figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)

dev.copy(png, figFile)
dev.off()

# --- specific --- #
allGenesS <- allGenes
DEGsS <- DEGs
```


### phyB.R vs WT.R

```{r echo=FALSE}
# --- inputs --- #
folder <- "phyB_v_WT_at_timepoint"
cont <- "phyB.R_vs_WT.R_exact"
conds <- c("phyB.R", "WT.R")
FDRmax <- 0.05
logFCmin <- 1
# --- ------ --- #
exTest <- exactTest(woundDGEdata, pair=conds)
allGenes <- as.data.frame(topTags(exTest, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addWoundCtrl(DEGs, woundIDs)
DEGs <- addAvgCpm(DEGs, DGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)

hist(allGenes$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")

figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)

dev.copy(png, figFile)
dev.off()

# --- specific --- #
allGenesR <- allGenes
DEGsR <- DEGs
```



## GT at all time points at once


### phyB vs WT at P,S,R

```{r}
# --- inputs --- #
folder <- "phyB_v_WT_at_timepoint"
cont <- "phyB.PSR_vs_WT.PSR"
conds <- c("WT.P", "phyB.P", "WT.S", "phyB.S", "WT.R", "phyB.R")
FDRmax <- 0.05
logFCmin <- 1
# --- ------ --- #

qlf <- glmQLFTest(LmFit, contrast=fullContrasts[, cont])
allGenes <- as.data.frame(topTags(qlf, n=Inf, p.value=1)) 
DEGs <- subset(allGenes, (FDR<FDRmax) & (abs(logFC)>logFCmin))
DEGs <- addWoundCtrl(DEGs, woundIDs)
DEGs <- addAvgCpm(DEGs, DGEdata, groups=conds)
save_DEGs(DEGs, folder, cont, FDRmax, logFCmin, dirPath)

hist(allGenes$PValue, breaks=40, main=paste(cont, "p-value histogram", sep=" "),
         xlab = "p-value")
figFile <- makeFname(folder=paste(folder, "/plots", sep=""), cont=cont, 
                     dtype="_p_val_hist", FDRmax=FDRmax, logFCmin=logFCmin)
figFile <- file.path(dirPath, figFile)
dev.copy(png, figFile)
dev.off()

# --- specific --- #


```






--------------------------------------------------------------------------------
# END MAIN ANALYSIS






```{r}
#plot.new()
#tripVenn(DEGsRvS, DEGsR, DEGsS, c("RvsS (WT+phyB)", "R (WTvs.phyB)", "S (WTvs.phyB)"))

nVenn(list("RvsS (WT+phyB)"=row.names(DEGsRvS), "R (WTvs.phyB)"=row.names(DEGsR),
           "S (WTvs.phyB)"=row.names(DEGsS)))

#plot.new()
#tripVenn(DEGsSvP, DEGsS, DEGsP, c("SvsP (WT+phyB)", "S (WTvs.phyB)", "P (WTvs.phyB)"))

nVenn(list("SvsP (WT+phyB)"=row.names(DEGsSvP), "S (WTvs.phyB)"=row.names(DEGsS),
           "P (WTvs.phyB)"=row.names(DEGsP)))

```

# PCA

```{r}

cpmData <- cpm(DGEdata)
cpmData <- cpmData[, !grepl("24$", colnames(cpmData))]
pcaData <- cpmData
  
pca <- prcomp(t(pcaData), scale=TRUE)



plotData <- as.data.frame(pca$x)
plotData$sample <- row.names(plotData)
plotData <- tidyr::separate(plotData, sample, into=c("genotype", "plant", "treatment"), sep="_", remove=FALSE)
plotData$group <- paste(plotData$genotype, plotData$treatment, sep="_")

Palette <- c("#68d53f", "#c7dd27", "#c6874e", "red")

ggplot(plotData, aes(x=PC1, y=PC2, color=treatment, shape=genotype)) + 
  geom_point(size=4) + scale_colour_manual(values=Palette)


```


```{r}
ggplot(plotData, aes(x=PC2, y=PC3, color=treatment, shape=genotype)) + 
  geom_point(size=4) + scale_colour_manual(values=Palette)
```
```{r}
theta = -0.33*pi
#theta = 0.4*pi

plotData$PC2.3A <- plotData$PC3*sin(theta) + plotData$PC2*cos(theta)
plotData$PC2.3B <- plotData$PC3*cos(theta) - plotData$PC2*sin(theta)

ggplot(plotData, aes(x=PC1, y=PC2.3A, color=treatment, shape=genotype)) + 
  geom_point(size=4) + scale_colour_manual(values=Palette)

newPCs <- data.frame("PC2.3A" = pca$rotation[, 3]*sin(theta) + 
                       pca$rotation[, 2]*cos(theta), 
                     "PC2.3B"=pca$rotation[, 3]*cos(theta) -
                       pca$rotation[, 2]*sin(theta), 
                     row.names=row.names(pca$rotation))

```




```{r}
PC2.3sdev <- data.frame("PC" = c(2,3), "sdev"=c(
  sqrt((pca$sdev[3]*sin(theta))^2 + (pca$sdev[2]*cos(theta))^2),
  sqrt((pca$sdev[3]*cos(theta))^2 + (pca$sdev[2]*sin(theta))^2)
))

sdevData <- data.frame("PC" = 1:length(pca$sdev), "sdev"=pca$sdev)





ggplot() + geom_point(data=sdevData, aes(x=PC, y=sdev, 
                                         colour="Principal Components"), size=4) + 
  geom_point(data=PC2.3sdev, aes(x=PC, y=sdev, color="PC2.3 rotation"), size=4) +
  labs(title="Standard deviation of principal components")

# plot(x=1:length(pca$sdev), y=pca$sdev)
```


### do a PCA with only genes listed as "Yes" for Leaf_development

```{r}
leafDevGenes <- subset(edinburghGeneInfo, Leaf_development=="YES")$Gene_ID
leafDevGenes <- leafDevGenes[leafDevGenes %in% row.names(pcaData)]

leafPcaData <- pcaData[leafDevGenes, ]
  
pca <- prcomp(t(leafPcaData), scale=TRUE)

plotData <- as.data.frame(pca$x)
plotData$sample <- row.names(plotData)
plotData <- tidyr::separate(plotData, sample, into=c("genotype", "plant#", "treatment"), sep="_", remove=FALSE)
plotData$group <- paste(plotData$genotype, plotData$treatment, sep="_")

Palette <- c("#68d53f", "#c7dd27", "#c6874e")

ggplot(plotData, aes(x=PC1, y=PC2, color=treatment, shape=genotype)) + 
  geom_point(size=4) + scale_colour_manual(values=Palette)


```

```{r}
ggplot(plotData, aes(x=PC2, y=PC3, color=treatment, shape=genotype)) + 
  geom_point(size=4) + scale_colour_manual(values=Palette)
```

```{r}
theta = -0.28*pi

plotData$PC2.3A <- plotData$PC3*sin(theta) + plotData$PC2*cos(theta)

ggplot(plotData, aes(x=PC1, y=PC2.3A, color=treatment, shape=genotype)) + 
  geom_point(size=4) + scale_colour_manual(values=Palette)

```





```{r}
plot(x=1:length(pca$sdev), y=pca$sdev)
```



